src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs

```diff
@@ -15,32 +16,71 @@ namespace System.Net.Http.Headers
     // Use HeaderDescriptor.TryGet to resolve an arbitrary header name to a HeaderDescriptor.
     internal readonly struct HeaderDescriptor : IEquatable<HeaderDescriptor>
     {
-        private readonly string _headerName;
-        private readonly KnownHeader? _knownHeader;
+        /// <summary>
+        /// Either a <see cref="KnownHeader"/> or <see cref="string"/>.
+        /// </summary>
+        private readonly object _descriptor;
 
         public HeaderDescriptor(KnownHeader knownHeader)
         {
-            _knownHeader = knownHeader;
-            _headerName = knownHeader.Name;
+            _descriptor = knownHeader;
         }
 
         // This should not be used directly; use static TryGet below
-        internal HeaderDescriptor(string headerName)
+        internal HeaderDescriptor(string headerName, bool customHeader = false)
         {
-            _headerName = headerName;
-            _knownHeader = null;
+            Debug.Assert(customHeader || KnownHeaders.TryGetKnownHeader(headerName) is null, $"The {nameof(KnownHeader)} overload should be used for {headerName}");
+            _descriptor = headerName;
         }
 
-        public string Name => _headerName;
-        public HttpHeaderParser? Parser => _knownHeader?.Parser;
-        public HttpHeaderType HeaderType => _knownHeader == null ? HttpHeaderType.Custom : _knownHeader.HeaderType;
-        public KnownHeader? KnownHeader => _knownHeader;
+        public string Name
+        {
+            get
+            {
+                object descriptor = _descriptor;
+                return descriptor.GetType() == typeof(KnownHeader) ?
+                    Unsafe.As<KnownHeader>(descriptor).Name :
+                    Unsafe.As<string>(descriptor);
+            }
+        }
+
+        public bool IsKnownHeader([NotNullWhen(true)] out KnownHeader? knownHeader, [NotNullWhen(false)] out string? headerName)
+        {
+            object descriptor = _descriptor;
+            if (descriptor.GetType() == typeof(KnownHeader))
+            {
+                knownHeader = Unsafe.As<KnownHeader>(descriptor);
```

Comment:
How is this different to `knownHeader = (Knownheader)descriptor;` ?

Reply:
The cast is more expensive (the JIT won't skip it based on the `GetType` check above). Overall it didn't make that big of a difference so I simplified the code.